In [1]:
import pandas as pd
import numpy as np

In [124]:
data = pd.read_csv('./tv_cleaned_v1.csv')
data.head()

Unnamed: 0  price  rating  screen_size    brand tv_size   tv_type  \
0           0  14999      53         32.0     acer    32in   QLED TV   
1           1   9999      53         43.0   xiaomi    43in    LED TV   
2           2  31990      65         43.0       lg    43in    LED TV   
3           3  38900      63         43.0  samsung    43in    LED TV   
4           4  16249      48         32.0       lg    32in    LED TV   

       resolution            pixels        hd  ...  has_ethernet num_speakers  \
0  1366x768pixels           missing        HD  ...           1.0            2   
1         Full HD   1920x1080pixels   Full HD  ...           1.0            2   
2   4K Resolution   3840x2160pixels  Ultra HD  ...           1.0            2   
3   4K Resolution   3840x2160pixels  Ultra HD  ...           1.0            2   
4  1366x768pixels           missing        HD  ...           1.0            2   

  speaker_output processor_core  processor_ram  processor_inbuilt_storage  \
0           30.0            4.0            1.5                       16.0   
1           16.0            NaN            1.0                        8.0   
2           20.0            NaN            NaN                        NaN   
3           20.0            NaN            NaN                        NaN   
4           16.0            NaN            NaN                        NaN   

  is_ultra_slim  viewing_angle  num_hdmi_ports  num_usb_ports  
0           1.0          178.0             2.0            2.0  
1           0.0          178.0             2.0            2.0  
2           NaN            NaN             3.0            2.0  
3           0.0          178.0             2.0            1.0  
4           0.0          178.0             3.0            2.0  

[5 rows x 25 columns]

In [125]:
data.drop(columns='Unnamed: 0', inplace=True)

# brand

In [126]:
data[data['brand'] == '16"9 Ratio']

price  rating  screen_size       brand tv_size tv_type resolution pixels  \
59  99999      45         27.0  16"9 Ratio      lg    27in    missing   Wifi   

      hd  is_smart  ... has_ethernet num_speakers speaker_output  \
59  27.0         1  ...          0.0          NaN            NaN   

    processor_core  processor_ram processor_inbuilt_storage  is_ultra_slim  \
59             NaN            NaN                       NaN            0.0   

    viewing_angle  num_hdmi_ports  num_usb_ports  
59          178.0             2.0            2.0  

[1 rows x 24 columns]

In [127]:
x = data.loc[59,'brand':'tv_type'].shift(-1).values 
data.loc[59,'brand':'tv_type'] = x

In [128]:
data['brand'] = data['brand'].str.strip().str.lower()

In [129]:
data.drop(columns='tv_size', inplace=True)

# tv type

In [130]:
data['tv_type'] = data['tv_type'].str.strip().str.lower()

In [23]:
for i in data['tv_type'].unique():
    print(i)

qled tv
led tv
microled tv
missing
nan
micro led tv
qned tv
oled tv
lcd tv
qd-led tv
mini-led tv
qd-mini led tv
mini led tv
laser tv
laser tv tv
mini led qled tv
qd-miniled tv
miniled tv
full array led tv


In [131]:
data.loc[data['tv_type'] == 'missing', 'tv_type'] = np.nan

In [62]:
for i in data['tv_type'].unique():
    print(i)

qled 
led 
micro led
nan
qned 
oled 
lcd 
qd-led 
mini led
qd mini led
laser 
laser  
mini led qled 
full array led 


In [132]:
data['tv_type'] = data['tv_type'].str.replace('tv','')
data.loc[data['tv_type'].fillna('missing').str.contains('micro'), 'tv_type'] = 'micro led'

In [133]:
filt = (data['tv_type'].str.contains('mini')) & (~data['tv_type'].fillna('missing').str.contains('qd')) & (~data['tv_type'].fillna('missing').str.contains('qled'))
data.loc[filt,'tv_type'] = 'mini led'

In [134]:
filt = (data['tv_type'].str.contains('mini')) & (data['tv_type'].fillna('missing').str.contains('led')) & (data['tv_type'].fillna('missing').str.contains('-'))
data.loc[filt, 'tv_type'] = 'qd mini led'

In [135]:
data['tv_type'].value_counts()

led                744
qled               161
oled                77
mini led            10
qd mini led          6
micro led            3
lcd                  2
qd-led               2
laser                2
laser                2
mini led qled        2
qned                 1
full array led       1
Name: tv_type, dtype: int64

### INFO ABOUT TV TYPE:

* LCDs accounted for the most of the TVs in the market but the term LED was coined a decade ago because the CCFL  flourescent lamps used in LCD screens were replaced with
light emitting diodes(LEDs) because they were brighter, lasted longer and allowed the screens to be flatter as compared to LCDs.


1. QLED: Quantum dot light emitting diode: it is like LED but has an additional quantum dot layer that enhances everything we see on screen like the colors and the brightness of the images.  All QLED TVs are also LCDs but they use quantum dots to produce colors

2. MiniLED : in these TVs, teh backlights are made up of mini LEDs for improved brightness and contrast.

3. OLED : organic light-emitting diode - in this each individual pixel emits its own light, so no additional backlight is required.


# resolution

In [69]:
data['resolution'].unique()

array(['1366x768pixels', 'Full HD', '4K Resolution', 'missing',
       '8K Resolution', '1366x728pixels', '7680x4380pixels',
       '1920x1083pixels', 'HD Ready', '366x768pixels', '1399x768pixels'],
      dtype=object)

In [70]:
data['pixels'].unique()

array(['missing', ' 1920x1080pixels', ' 3840x2160pixels',
       '1366x768pixels', ' 7680x4320pixels', 'Wifi', ' 1080x1920pixels',
       ' 2160x3840pixels', ' 3810x2160pixels', ' 3480x2160pixels',
       ' 1280x720pixels', ' 1260x720pixels', ' 3860x2160pixels',
       ' 840x2160pixels', ' 9840x2160pixels'], dtype=object)

In [136]:
temp = data[data['resolution'].str.contains('pixels')]
data.loc[temp.index, 'pixels'] = data.loc[temp.index, 'resolution']
data.loc[temp.index, 'resolution'] = np.nan

In [75]:
data['resolution'].unique()

array([nan, 'Full HD', '4K Resolution', 'missing', '8K Resolution',
       'HD Ready'], dtype=object)

In [137]:
data.loc[data['resolution'] == 'missing', 'resolution'] = np.nan

In [138]:
data['resolution'].nunique()

4

# pixels
if pixels = 1366x768 -> x-resolution orwidth or horizontal resolution=1366 and y-resolution or height or vertical resolution = 768

In [81]:
data['pixels'].unique()

array(['1366x768pixels', ' 1920x1080pixels', ' 3840x2160pixels',
       ' 7680x4320pixels', '1366x728pixels', 'Wifi', ' 1080x1920pixels',
       '7680x4380pixels', '1920x1083pixels', ' 2160x3840pixels',
       ' 3810x2160pixels', ' 3480x2160pixels', ' 1280x720pixels',
       ' 1260x720pixels', ' 3860x2160pixels', ' 840x2160pixels',
       '366x768pixels', '1399x768pixels', ' 9840x2160pixels'],
      dtype=object)

In [139]:
data['pixels'] = data['pixels'].str.strip()

In [140]:
data[data['pixels'] == 'Wifi']
data.loc[59,'pixels'] = np.nan

In [141]:
data[['x_resolution','y_resolution']] =data['pixels'].str.replace('pixels','').str.split('x', expand=True)

In [142]:
data[['pixels', 'x_resolution', 'y_resolution']].sample(7)

pixels x_resolution y_resolution
977   1920x1080pixels         1920         1080
389   1920x1080pixels         1920         1080
508   1920x1080pixels         1920         1080
636   3840x2160pixels         3840         2160
976   3840x2160pixels         3840         2160
1004   1366x768pixels         1366          768
825   3840x2160pixels         3840         2160

In [144]:
data.drop(columns='pixels', inplace=True)

# hd

In [105]:
data['hd'].unique()

array(['HD', 'Full HD', 'Ultra HD', 'missing', '27.0'], dtype=object)

In [145]:
data[data['hd'] == '27.0']
data.loc[59,'hd'] = np.nan

In [146]:
data.loc[data['hd']=='missing'] = np.nan

In [147]:
data['hd'].unique()

array(['HD', 'Full HD', 'Ultra HD', nan], dtype=object)

# is smart

In [148]:
data['is_smart'].unique()

array([ 1., nan,  0.])

# led type

In [149]:
data['led_type'].unique()

array(['QLED', 'LED', 'MicroLED', nan, '0', 'missing'], dtype=object)

In [150]:
data.loc[data['led_type'] == 'missing', 'led_type'] = np.nan

In [152]:
data['led_type'].unique()

array(['QLED', 'LED', 'MicroLED', nan, '0'], dtype=object)

In [154]:
data[data['led_type'] == '0']
data.loc[59,'led_type']  = np.nan

In [157]:
data['led_type'] = data['led_type'].str.strip().str.lower()

In [172]:
data[['tv_type', 'led_type']].sample(7)

tv_type led_type
409    led       led
779    led       led
9      led       led
717    led       led
743    led       led
786    led       led
906     NaN      NaN

In [177]:
data.loc[274,'tv_type'] = 'qd led'
data.loc[440,'tv_type'] = 'qd mini led'
data.loc[441,'tv_type'] = 'qd mini led'
data.loc[442,'tv_type'] = 'qd mini led'
data.loc[671,'tv_type'] = 'qd mini led'

In [178]:
data.loc[672,'tv_type'] = 'qd led'
data.loc[673,'tv_type'] = 'qd mini led'
data.loc[674,'tv_type'] = 'qd mini led'

In [179]:
data[data['tv_type'].fillna('missing').str.contains('qd')]

price  rating  screen_size brand      tv_type     resolution  \
274  310999.0    73.0         98.0   tcl       qd led  4K Resolution   
440  194499.0    62.0         85.0   tcl  qd mini led  4K Resolution   
441  102999.0    60.0         65.0   tcl  qd mini led  4K Resolution   
442  137299.0    61.0         75.0   tcl  qd mini led  4K Resolution   
671  599999.0    75.0         98.0   tcl  qd mini led  4K Resolution   
672   95999.0    69.0         65.0   tcl       qd led  4K Resolution   
673  287999.0    73.0         75.0   tcl  qd mini led  4K Resolution   
674  383999.0    74.0         85.0   tcl  qd mini led  4K Resolution   

           hd  is_smart led_type resolution_type  ... speaker_output  \
274  Ultra HD       1.0      led   4K Resolution  ...           60.0   
440  Ultra HD       1.0      led   4K Resolution  ...           90.0   
441  Ultra HD       1.0      led   4K Resolution  ...           90.0   
442  Ultra HD       1.0      led   4K Resolution  ...           90.0   
671  Ultra HD       1.0      led   4K Resolution  ...           60.0   
672  Ultra HD       1.0      led   4K Resolution  ...           60.0   
673  Ultra HD       1.0      led   4K Resolution  ...           60.0   
674  Ultra HD       1.0      led   4K Resolution  ...           60.0   

     processor_core  processor_ram processor_inbuilt_storage  is_ultra_slim  \
274             NaN            4.0                      64.0            NaN   
440             NaN            4.0                      64.0            0.0   
441             NaN            4.0                      64.0            0.0   
442             NaN            4.0                      64.0            0.0   
671             NaN            4.0                     128.0            NaN   
672             NaN            4.0                      64.0            NaN   
673             NaN            4.0                     128.0            NaN   
674             NaN            4.0                     128.0            NaN   

     viewing_angle  num_hdmi_ports  num_usb_ports  x_resolution  y_resolution  
274            NaN             4.0            4.0          3840          2160  
440            NaN             NaN            NaN          3840          2160  
441            NaN             NaN            NaN          3840          2160  
442            NaN             NaN            NaN          3840          2160  
671            NaN             4.0            4.0          3840          2160  
672            NaN             4.0            4.0          3840          2160  
673            NaN             4.0            4.0          3840          2160  
674            NaN             4.0            4.0          3840          2160  

[8 rows x 24 columns]

In [176]:
data['led_type'].isna().mean()

0.1588235294117647

In [182]:
data['tv_type'].unique()

array(['qled ', 'led ', 'micro led', nan, 'qned ', 'oled ', 'lcd ',
       'qd led', 'mini led', 'qd mini led', 'laser ', 'laser  ',
       'mini led qled ', 'full array led '], dtype=object)

In [185]:
temp = data[data['tv_type'].fillna('missing').str.contains('laser')]
data.loc[temp.index, 'tv_type'] = 'laser'

In [202]:
data['tv_type'].unique()

array(['qled', 'led', 'micro led', nan, 'qned', 'oled', 'lcd', 'qd led',
       'mini led', 'qd mini led', 'laser', 'mini led qled',
       'full array led'], dtype=object)

In [188]:
temp = data[data['tv_type'].fillna('missing').str.contains('lcd')]
data.loc[temp.index,'tv_type'] = 'lcd'

In [198]:
pd.set_option('display.max_rows', None)
import warnings
warnings.filterwarnings('ignore')

In [199]:
temp = data[data['tv_type'].fillna('missing').str.contains('qled')]
temp.drop(index=[574,575], inplace=True)
data.loc[temp.index, 'tv_type'] = 'qled'

In [ ]:
data[data['tv_type'].fillna('missing').str.contains('qled')]

In [201]:
data['tv_type'] = data['tv_type'].str.strip()

- tv type and led_type have the same data.. the kind of led on the tv screen.
- but since, tv_type has only 7% null values as compared to led_type which has 16% null values, I am dropping led_type

In [203]:
data['tv_type'].isna().mean()*100

6.96078431372549

In [204]:
data['led_type'].isna().mean()*100

15.88235294117647

In [205]:
data.drop(columns='led_type', inplace=True)

In [206]:
data.columns

Index(['price', 'rating', 'screen_size', 'brand', 'tv_type', 'resolution',
       'hd', 'is_smart', 'resolution_type', 'screen_pixels', 'has_wifi',
       'has_ethernet', 'num_speakers', 'speaker_output', 'processor_core',
       'processor_ram', 'processor_inbuilt_storage', 'is_ultra_slim',
       'viewing_angle', 'num_hdmi_ports', 'num_usb_ports', 'x_resolution',
       'y_resolution'],
      dtype='object')

In [211]:
data.drop(columns='screen_pixels', inplace=True)

# has_wifi

In [212]:
data['has_wifi'].unique()

array([ 1., nan,  0.])

# has ethernet

In [214]:
data['has_ethernet'].unique()

array([ 1.,  0., nan])

# num_speakers

In [216]:
data['num_speakers'].unique()

array(['2', '4', nan, '5', '3', '7', '1', '50', '6', '8', '4.2.2'],
      dtype=object)

In [218]:
data.loc[data['num_speakers'] == '4.2.2', 'num_speakers'] = 8

In [219]:
data['num_speakers'] = data['num_speakers'].astype(float)

# speaker output

In [222]:
data['speaker_output'].dtype

dtype('float64')

# processor core

In [225]:
data['processor_core'].unique()

array([ 4., nan,  8.,  6.])

# processor inbuilt storage

In [226]:
data['processor_inbuilt_storage'].unique()

array([ 16. ,   8. ,   nan,  32. ,  64. ,   4. ,   0.5,   2. , 128. ])

# is ultra slim

In [227]:
data['is_ultra_slim'].unique()

array([ 1.,  0., nan])

# viewing angle

In [228]:
data['viewing_angle'].unique()

array([178.,  nan, 180., 179.,   8., 170.])

In [229]:
data.columns

Index(['price', 'rating', 'screen_size', 'brand', 'tv_type', 'resolution',
       'hd', 'is_smart', 'resolution_type', 'has_wifi', 'has_ethernet',
       'num_speakers', 'speaker_output', 'processor_core', 'processor_ram',
       'processor_inbuilt_storage', 'is_ultra_slim', 'viewing_angle',
       'num_hdmi_ports', 'num_usb_ports', 'x_resolution', 'y_resolution'],
      dtype='object')

# processor ram

In [230]:
data['processor_ram'].unique()

array([1.5 , 1.  ,  nan, 2.  , 3.  , 4.  , 0.5 , 2.5 , 1.75, 6.  , 8.  ])

# num_hdmi_ports

In [231]:
data['num_hdmi_ports'].unique()

array([ 2.,  3.,  1., nan,  4.,  6.])

# num_usb_ports

In [232]:
data['num_usb_ports'].unique()

array([ 2.,  1., nan,  3.,  4.])

In [235]:
for i in data.columns:
    print(i)

price
rating
screen_size
brand
tv_type
resolution
hd
is_smart
resolution_type
has_wifi
has_ethernet
num_speakers
speaker_output
processor_core
processor_ram
processor_inbuilt_storage
is_ultra_slim
viewing_angle
num_hdmi_ports
num_usb_ports
x_resolution
y_resolution


In [236]:
data  = (data[['brand','tv_type', 'resolution', 'resolution_type', 'x_resolution','y_resolution', 'screen_size', 'processor_core', 'processor_ram', 'processor_inbuilt_storage',
      'num_speakers', 'speaker_output', 'hd', 'has_wifi', 'has_ethernet', 'is_smart', 'is_ultra_slim', 'viewing_angle', 'num_hdmi_ports', 'num_usb_ports', 'rating',
      'price']])

In [237]:
data.head()

brand tv_type     resolution resolution_type x_resolution y_resolution  \
0     acer    qled            NaN         missing         1366          768   
1   xiaomi     led        Full HD         missing         1920         1080   
2       lg     led  4K Resolution   4K Resolution         3840         2160   
3  samsung     led  4K Resolution   4K Resolution         3840         2160   
4       lg     led            NaN         missing         1366          768   

   screen_size  processor_core  processor_ram  processor_inbuilt_storage  ...  \
0         32.0             4.0            1.5                       16.0  ...   
1         43.0             NaN            1.0                        8.0  ...   
2         43.0             NaN            NaN                        NaN  ...   
3         43.0             NaN            NaN                        NaN  ...   
4         32.0             NaN            NaN                        NaN  ...   

         hd  has_wifi has_ethernet  is_smart  is_ultra_slim  viewing_angle  \
0        HD       1.0          1.0       1.0            1.0          178.0   
1   Full HD       1.0          1.0       1.0            0.0          178.0   
2  Ultra HD       1.0          1.0       1.0            NaN            NaN   
3  Ultra HD       1.0          1.0       1.0            0.0          178.0   
4        HD       1.0          1.0       1.0            0.0          178.0   

   num_hdmi_ports  num_usb_ports  rating    price  
0             2.0            2.0    53.0  14999.0  
1             2.0            2.0    53.0   9999.0  
2             3.0            2.0    65.0  31990.0  
3             2.0            1.0    63.0  38900.0  
4             3.0            2.0    48.0  16249.0  

[5 rows x 22 columns]

# importing the data

In [238]:
data.to_csv('tv_cleaned_v2.csv')